CI: Add Latest Intel Compiler

ax3l opened the pull request:
Add latest `icpc`/`icc` from oneAPI for Linux to CI.

Related to #995

Since it's currently plain impossible to install boost with icc on Linux, we install the header-only variant of boost with GCC and use that (no Boost.Fibers coverage in CI).

ax3l committed:
CI: Add Latest Intel Compiler

Add latest `icpc`/`icc` to CI.

ax3l commented:
Come on Boost, urgh...
```
+ sudo ./bootstrap.sh --with-toolset=intel-linux
Building Boost.Build engine with toolset intel-linux... 
Failed to build Boost.Build build engine
Consult 'bootstrap.log' for more details
```
(the outdated syntax `--with-toolset=intel` gives the same error)

ax3l committed:
Boost + Intel: Show logfile

ax3l committed:
Intel: Check CC after install

ax3l commented:
```
which "${CC}"
++ which icc
${CC} -v
/opt/intel/inteloneapi/compiler/latest/linux/bin/intel64/icc
++ icc -v
icc version 2021.1 Beta (gcc version 7.5.0 compatibility)
...
icc -o bootstrap/jam0 command.c compile.c constants.c debug.c execcmd.c frames.c function.c glob.c hash.c hdrmacro.c headers.c jam.c jambase.c jamgram.c lists.c make.c make1.c object.c option.c output.c parse.c pathsys.c regexp.c rules.c scan.c search.c subst.c timestamp.c variable.c modules.c strings.c filesys.c builtins.c class.c cwd.c native.c md5.c w32_getreg.c modules/set.c modules/path.c modules/regex.c modules/property-set.c modules/sequence.c modules/order.c execunix.c fileunix.c pathunix.c
./build.sh: 17: ./build.sh: icc: not found
```

BenjaminW3 commented on script/install_icpc.sh:
```diff
@@ -0,0 +1,39 @@
+which "${CXX}"
+${CXX} -v
+which "${CC}"
+${CC} -v
```
Could you try to add those 4 lines in front of the `./bootstrap.sh --with-toolset="intel-linux"`.
Maybe some environment vars get lost in between?

ax3l replied:
I think so, yep.

BenjaminW3 replied:
It may be necessary that you have to call `source /opt/intel/inteloneapi/setvars.sh` in `alpaka/script/run.sh` again.
I have not yet removed the split between the docker image creation (start, run install.sh, stop, save as image) and then building and executing the tests (start image, run.sh to build and test). The environment is not correctly preserved in between those steps.
I already started getting rid of this two step setup because we are not caching the docker images anymore with github actions.

ax3l replied:
Looks like the `icc` is there. something funky in bootstrap of bjam is going on.

ax3l replied:
Yep, the boost build system is archaic:
https://github.com/boostorg/build/issues/475
https://github.com/boostorg/build/pull/635
```
A C++11 capable compiler is required for building the B2 engine.
Toolset 'intel-linux' does not appear to support C++11
```
This does not even make sense, since `icc` already defaults to C++14... The compile feature check is super outdated.

I wonder how far the CMake adoption process in Boost is in 2020...
Update: aha
```
CMake Error at tools/cmake/include/BoostRoot.cmake:13 (message):
  CMake support in Boost is experimental and part of an ongoing development
  effort.  It's not ready for use yet.  Please use b2 (Boost.Build) to build
  and install Boost.
```

ax3l committed:
Boost Intel: Debug further

ax3l committed:
Intel: Use Latest Boost

https://stackoverflow.com/a/59255999/2719194

ax3l committed:
Boost: cxx toolset

This special toolset reacts on CXX, CXXFLAGS et al.

ax3l commented:
The boost build system is archaic:
https://github.com/boostorg/build/issues/475

ax3l committed:
Boost CXXFLAGS: C++11

ax3l committed:
Put std into CXX

ax3l committed:
ICC defaults should do with cxx toolset

ax3l committed:
Intel Boost: Just patch the new flags in

ax3l committed:
Boost: Try new CMake scripts...

ax3l committed:
Boost Intel: Try a bugfix PR

ax3l commented:
@krzikalla how do you install boost with the Intel compilers on Linux?

ax3l committed:
Boost Intel: Install via GCC

Totally fine as long as we only use header-only libs.

ax3l committed:
CI: Source Install

Try to preserve compiler activation env.

ax3l commented:
@BenjaminW3 I am now successfully installing Boost with GCC for this ICC build.
But we might have to preserve the activated compiler environment better or source `source /opt/intel/oneapi/setvars.sh` again in the build.

ax3l committed:
icpc: re-init inside docker

Too many vars are set in the install phase of oneAPI.

ax3l committed:
oneAPI: Add mkl, openmp

```
ld: cannot find /opt/intel/oneapi/compiler/2021.1-beta08/linux/bin/intel64/../../compiler/lib/intel64_lin/i_ofldbegin_target.o: No such file or directory
```
mkl: no change
openmp: no change
tbb: ... let's see.

BenjaminW3 commented on .github/workflows/ci.yml:
```diff
@@ -214,6 +214,11 @@ jobs:
+        # icpc
+        - name: linux_icpc_release
+          os: ubuntu-latest
+          env: {TRAVIS_OS_NAME: linux,   CXX: icpc,    CC: icc, CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.73.0, ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_CI_CMAKE_VER: 3.15.7, OMP_NUM_THREADS: 2, ALPAKA_CI_STDLIB: libstdc++, ALPAKA_CI_DOCKER_BASE_IMAGE_NAME: "ubuntu:18.04"}
```
`TRAVIS_OS_NAME` is not necessary anymore.
It has been renamed to `ALPAKA_CI_OS_NAME` and is now automatically added since #1089

BenjaminW3 replied:
The order of env variables has been changed and `ALPAKA_CI_STDLIB` has been moved to behind `CC`.

BenjaminW3 commented on script/ci.sh:
```diff
@@ -17,7 +17,7 @@ source ./script/before_install.sh
 if [ "$ALPAKA_CI_OS_NAME" = "Linux" ]
 then
-  ./script/docker_install.sh
+  source ./script/docker_install.sh
```
I do not think that this works ;-)

BenjaminW3 replied:
With #1109 install.sh and run.sh will be run within a single invocation of `docker run` which should make it much easier to keep settings between the invocations.

BenjaminW3 commented on script/install_icpc.sh:
```diff
@@ -0,0 +1,39 @@
+
+source ./script/set.sh
+
+: "${CXX?'CXX must be specified'}"
```
You also require `CC`.

BenjaminW3 commented on .github/workflows/ci.yml:
```diff
@@ -214,6 +214,11 @@ jobs:
+        # icpc
+        - name: linux_icpc_release
+          os: ubuntu-latest
+          env: {TRAVIS_OS_NAME: linux,   CXX: icpc,    CC: icc, CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.73.0, ALPAKA_ACC_CPU_B_SEQ_T_FIBERS_ENABLE: OFF, ALPAKA_CI_CMAKE_VER: 3.15.7, OMP_NUM_THREADS: 2, ALPAKA_CI_STDLIB: libstdc++, ALPAKA_CI_DOCKER_BASE_IMAGE_NAME: "ubuntu:18.04"}
```
May we have to disable the fibers backend because we would be mixing gcc built boost libs with an intel compiler built executable?

ax3l commented:
Didn't I disabled that already? Please feel free to push on the PR's branch, I am off now for two weeks. 🌴

BenjaminW3 commented:
Merged via #1235
